apps/emqx_ft/src/emqx_ft_storage_exporter_fs.erl

```diff
@@ -0,0 +1,333 @@
+                (_Path, Fileinfo = #file_info{type = regular}, [Filename | _], Acc) ->
+                    Filepath = filename:join(mk_result_dirname(Transfer) ++ [Filename]),
+                    Info = mk_exportinfo(Options, Filename, Filepath, Transfer, Fileinfo),
+                    [Info | Acc];
+                (_Path, Fileinfo = #file_info{}, _, Acc) ->
+                    Acc;
+                (_Path, {error, Reason}, _, Acc) ->
+                    Acc
+            end,
+            [],
+            Dirpath,
+            [fun filter_manifest/1]
+        )
+    of
+        Infos = [_ | _] ->
+            {ok, Infos};
+        [] ->
+            {error, enoent};
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+-spec list_local(options()) ->
+    {ok, #{transfer() => [exportinfo(), ...]}}.
+list_local(Options) ->
+    Pattern = [
+        _Bucket1 = '*',
+        _Bucket2 = '*',
+        _Rest = '*',
+        _ClientId = '*',
+        _FileId = '*',
+        fun filter_manifest/1
+    ],
+    Root = get_storage_root(Options),
+    {ok,
+        emqx_ft_fs_util:fold(
+            fun(Filepath, Info, Stack, Acc) ->
+                read_exportinfo(Options, Filepath, Info, Stack, Acc)
+            end,
+            [],
+            Root,
+            Pattern
+        )}.
+
+filter_manifest(?MANIFEST) ->
+    % Filename equals `?MANIFEST`, there should also be a manifest for it.
+    false;
+filter_manifest(Filename) ->
+    ?MANIFEST =/= string:find(Filename, ?MANIFEST, trailing).
+
+read_exportinfo(Options, Filepath, Fileinfo = #file_info{type = regular}, Stack, Acc) ->
+    [Filename, FileId, ClientId | _] = Stack,
+    Transfer = {iolist_to_binary(ClientId), iolist_to_binary(FileId)},
+    Info = mk_exportinfo(Options, Filename, Filepath, Transfer, Fileinfo),
+    [Info | Acc];
+read_exportinfo(_Options, Filepath, Fileinfo = #file_info{}, _Stack, Acc) ->
+    % TODO worth logging?
+    Acc;
+read_exportinfo(_Options, Filepath, {error, Reason}, _Stack, Acc) ->
+    % TODO worth logging?
+    Acc.
+
+mk_exportinfo(Options, Filename, Filepath, Transfer, Fileinfo) ->
+    Root = get_storage_root(Options),
+    try_read_filemeta(
+        filename:join(Root, mk_manifest_filepath(Filepath)),
+        #{
+            transfer => Transfer,
+            name => Filename,
+            uri => mk_export_uri(Options, Filepath),
+            timestamp => Fileinfo#file_info.mtime,
+            size => Fileinfo#file_info.size,
+            path => filename:join(Root, Filepath)
+        }
+    ).
+
+try_read_filemeta(Filepath, Info) ->
+    case emqx_ft_fs_util:read_decode_file(Filepath, fun decode_filemeta/1) of
+        {ok, Filemeta} ->
+            Info#{meta => Filemeta};
+        {error, Reason} ->
+            % TODO worth logging?
+            Info
+    end.
+
+mk_export_uri(Options, Filepath) ->
+    % emqx_ft_storage_exporter_fs_api:mk_export_uri(Options, Filepath).
+    emqx_ft_api:mk_file_uri(Options, node(), Filepath).
```

Comment:
M.b. return something like `{node(), Filepath}` and decorate it as an URL in the controller?
Or inject an "url-maker" through `Options` 🤔  

Otherwise we have a hard dependency in the deep internals from a peripheral module 😓

Reply:
My idea was that we'd use the code from the commented line here, depending on `emqx_ft_storage_exporter_fs_api` module (which is obviously non-existent, though I tried to hint the need for it in [this comment](https://github.com/keynslug/emqx/blob/87e423ee3f8ed21b30ff80ee52a14111b330eaab/apps/emqx_ft/src/emqx_ft_api.erl#L76-L80)), which to me sounds find from this point of view.